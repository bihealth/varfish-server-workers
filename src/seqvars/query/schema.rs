//! Supporting code for seqvar query definition.

use noodles_vcf as vcf;
use strum::IntoEnumIterator;

/// Variant effects.
#[derive(
    serde::Serialize,
    serde::Deserialize,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Debug,
    Clone,
    Copy,
    strum::EnumIter,
)]
pub enum VariantEffect {
    /// 3' UTR exon variant.
    #[serde(rename = "3_prime_UTR_exon_variant")]
    ThreePrimeUtrExonVariant,
    /// 3' UTR intron variant.
    #[serde(rename = "3_prime_UTR_intron_variant")]
    ThreePrimeUtrIntronVariant,
    /// 5' UTR exon variant.
    #[serde(rename = "5_prime_UTR_exon_variant")]
    FivePrimeUtrExonVariant,
    /// 5' UTR intron variant.
    #[serde(rename = "5_prime_UTR_intron_variant")]
    FivePrimeUtrIntronVariant,
    /// Coding transcript intron variant.
    #[serde(rename = "coding_transcript_intron_variant")]
    CodingTranscriptIntronVariant,
    /// Complex substitution.
    #[serde(rename = "complex_substitution")]
    ComplexSubstitution,
    /// Direct tandem duplication.
    #[serde(rename = "direct_tandem_duplication")]
    DirectTandemDuplication,
    /// Disruptive in-frame deletion.
    #[serde(rename = "disruptive_inframe_deletion")]
    DisruptiveInframeDeletion,
    /// Disruptive in-frame insertion.
    #[serde(rename = "disruptive_inframe_insertion")]
    DisruptiveInframeInsertion,
    /// Downstream gene variant.
    #[serde(rename = "downstream_gene_variant")]
    DownstreamGeneVariant,
    /// Exon loss variant.
    #[serde(rename = "exon_loss_variant")]
    ExonLossVariant,
    /// Feature truncation.
    #[serde(rename = "feature_truncation")]
    FeatureTruncation,
    /// Frameshift elongation.
    #[serde(rename = "frameshift_elongation")]
    FrameshiftElongation,
    /// Frameshift truncation.
    #[serde(rename = "frameshift_truncation")]
    FrameshiftTruncation,
    /// Frameshift variant.
    #[serde(rename = "frameshift_variant")]
    FrameshiftVariant,
    /// In-frame deletion.
    #[serde(rename = "inframe_deletion")]
    InframeDeletion,
    /// In-frame insertion.
    #[serde(rename = "inframe_insertion")]
    InframeInsertion,
    /// Intergenic variant.
    #[serde(rename = "intergenic_variant")]
    IntergenicVariant,
    /// Internal feature elongation.
    #[serde(rename = "internal_feature_elongation")]
    InternalFeatureElongation,
    /// Missense variant.
    #[serde(rename = "missense_variant")]
    MissenseVariant,
    /// MNV.
    #[serde(rename = "mnv")]
    Mnv,
    /// Non-coding transcript exon variant.
    #[serde(rename = "non_coding_transcript_exon_variant")]
    NonCodingTranscriptExonVariant,
    /// Non-coding transcript intron variant.
    #[serde(rename = "non_coding_transcript_intron_variant")]
    NonCodingTranscriptIntronVariant,
    /// Splice acceptor variant.
    #[serde(rename = "splice_acceptor_variant")]
    SpliceAcceptorVariant,
    /// Splice donor variant.
    #[serde(rename = "splice_donor_variant")]
    SpliceDonorVariant,
    /// Splice region variant.
    #[serde(rename = "splice_region_variant")]
    SpliceRegionVariant,
    /// Start lost.
    #[serde(rename = "start_lost")]
    StartLost,
    /// Stop gained.
    #[serde(rename = "stop_gained")]
    StopGained,
    /// Stop lost.
    #[serde(rename = "stop_lost")]
    StopLost,
    /// Stop retained variant.
    #[serde(rename = "stop_retained_variant")]
    StopRetainedVariant,
    /// Structural variant.
    #[serde(rename = "structural_variant")]
    StructuralVariant,
    /// Synonymous variant.
    #[serde(rename = "synonymous_variant")]
    SynonymousVariant,
    /// Transcript ablation.
    #[serde(rename = "transcript_ablation")]
    TranscriptAblation,
    /// Upstream gene variant.
    #[serde(rename = "upstream_gene_variant")]
    UpstreamGeneVariant,
}

impl VariantEffect {
    /// Return vector of all values of `VariantEffect`.
    pub fn all() -> Vec<Self> {
        Self::iter().collect()
    }
}

/// Enumeration for recessive mode queries.
#[derive(
    serde::Serialize, serde::Deserialize, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Clone, Copy,
)]
pub enum RecessiveMode {
    /// Recessive.
    #[serde(rename = "recessive")]
    Recessive,
    /// Compound recessive.
    #[serde(rename = "compound-recessive")]
    CompoundRecessive,
}

/// Choices for failing quality thresholds on genotypes.
#[derive(
    serde::Serialize,
    serde::Deserialize,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Debug,
    Clone,
    Copy,
    Default,
)]
pub enum FailChoice {
    /// Ignore failure.
    #[default]
    #[serde(rename = "ignore")]
    Ignore,
    /// Drop whole variant.
    #[serde(rename = "drop-variant")]
    Drop,
    /// Interpret as "no-call".
    #[serde(rename = "no-call")]
    NoCall,
}

/// Choice for genotype.
#[derive(
    serde::Serialize,
    serde::Deserialize,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Debug,
    Clone,
    Copy,
    Default,
)]
pub enum GenotypeChoice {
    /// Any genotype.
    #[default]
    #[serde(rename = "any")]
    Any,
    /// Ref. genotype.
    #[serde(rename = "ref")]
    Ref,
    /// Het. genotype.
    #[serde(rename = "het")]
    Het,
    /// Hom. genotype.
    #[serde(rename = "hom")]
    Hom,
    /// Non-hom. genotype.
    #[serde(rename = "non-hom")]
    NonHom,
    /// Variant genotype.
    #[serde(rename = "variant")]
    Variant,
    /// Non-variant genotype.
    #[serde(rename = "non-variant")]
    NonVariant,
    /// Non-reference genotype.
    #[serde(rename = "non-reference")]
    NonReference,
    /// Index in comp. het. recessive inheritance.
    #[serde(rename = "comphet-index")]
    ComphetIndex,
    /// Index in recessive inheritance.
    #[serde(rename = "recessive-index")]
    RecessiveIndex,
    /// Parent in recessive inheritance.
    #[serde(rename = "recessive-parent")]
    RecessiveParent,
}

/// Quality settings for one sample.
#[derive(serde::Serialize, serde::Deserialize, PartialEq, Debug, Clone, Default)]
pub struct QualitySettings {
    /// Minimal coverage for het. sites.
    pub dp_het: Option<i32>,
    /// Minimal coverage for hom. sites.
    pub dp_hom: Option<i32>,
    /// Minimal genotype quality.
    pub gq: Option<i32>,
    /// Minimal allele balance.
    pub ab: Option<f32>,
    /// Minimal number of alternate reads.
    pub ad: Option<i32>,
    /// Maximal number of alternate reads
    pub ad_max: Option<i32>,
    /// Behaviour on failing quality thresholds.
    pub fail: FailChoice,
}

/// Data structure to hold a range.
#[derive(
    serde::Serialize, serde::Deserialize, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Clone,
)]
pub struct Range {
    /// Start of range.
    pub start: i32,
    /// End of range.
    pub end: i32,
}

/// Data struture to hold a genomic region.
#[derive(
    serde::Serialize, serde::Deserialize, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Clone,
)]
pub struct GenomicRegion {
    /// Chromosome.
    pub chrom: String,
    /// Range of region.
    pub range: Option<Range>,
}

/// Data structure with a single query.
#[derive(serde::Serialize, serde::Deserialize, PartialEq, Debug, Clone)]
#[serde(default)]
pub struct CaseQuery {
    /// Effects to consider.
    pub effects: Vec<VariantEffect>,

    /// Whether to enable filtration by gnomAD exomes.
    pub gnomad_exomes_enabled: bool,
    /// Whether to enable filtration by gnomAD genomes
    pub gnomad_genomes_enabled: bool,
    /// Whether to enable filtration by 1000 Genomes.
    pub inhouse_enabled: bool,
    /// Whether to enable filtration by mtDB.
    pub helixmtdb_enabled: bool,

    /// Quality settings for each individual.
    pub quality: indexmap::IndexMap<String, QualitySettings>,
    /// Genotype choice for each individual.
    pub genotype: indexmap::IndexMap<String, Option<GenotypeChoice>>,
    /// List of selected variants.
    pub selected_variants: Option<Vec<String>>, // TODO: remove?

    /// Whether to include coding transcripts.
    pub transcripts_coding: bool,
    /// Whether to include non-coding transcripts.
    pub transcripts_noncoding: bool,

    /// Whether to include SNVs.
    pub var_type_snv: bool,
    /// Whether to include indels.
    pub var_type_indel: bool,
    /// Whether to include MNVs.
    pub var_type_mnv: bool,

    /// Maximal distance to next exon, if any.
    pub max_exon_dist: Option<i32>,

    /// List of HGNC symbols, HGNC:<ID>s, ENSG<ID>s, or NCBI Gene IDs to restrict
    /// the resulting variants to.
    pub gene_allowlist: Option<Vec<String>>,
    /// List of genomic regions to limit restrict the resulting variants to.
    pub genomic_regions: Option<Vec<GenomicRegion>>,

    /// Wether to require ClinVar membership.
    pub require_in_clinvar: bool,
    /// Whether to include benign ClinVar variants.
    pub clinvar_include_benign: bool,
    /// Whether to include pathogenic ClinVar variants.
    pub clinvar_include_pathogenic: bool,
    /// Whether to include likely benign ClinVar variants.
    pub clinvar_include_likely_benign: bool,
    /// Whether to include likely pathogenic ClinVar variants.
    pub clinvar_include_likely_pathogenic: bool,
    /// Whether to include uncertain significance ClinVar variants.
    pub clinvar_include_uncertain_significance: bool,

    /// Maximal frequency in gnomAD exomes.
    pub gnomad_exomes_frequency: Option<f32>,
    /// Maximal number of heterozygous carriers in gnomAD exomes.
    pub gnomad_exomes_heterozygous: Option<i32>,
    /// Maximal number of homozygous carriers in gnomAD exomes.
    pub gnomad_exomes_homozygous: Option<i32>,
    /// Maximal number of hemizygous carriers in gnomAD exomes.
    pub gnomad_exomes_hemizygous: Option<i32>,

    /// Maximal frequency in gnomAD genomes.
    pub gnomad_genomes_frequency: Option<f32>,
    /// Maximal number of heterozygous carriers in gnomAD genomes.
    pub gnomad_genomes_heterozygous: Option<i32>,
    /// Maximal number of homozygous carriers in gnomAD genomes.
    pub gnomad_genomes_homozygous: Option<i32>,
    /// Maximal number of hemizygous carriers in gnomAD genomes.
    pub gnomad_genomes_hemizygous: Option<i32>,

    /// Maximal number of in-house carriers.
    pub inhouse_carriers: Option<i32>,
    /// Maximal number of in-house heterozygous carriers.
    pub inhouse_heterozygous: Option<i32>,
    /// Maximal number of in-house homozygous carriers.
    pub inhouse_homozygous: Option<i32>,
    /// Maximal number of in-house hemizygous carriers.
    pub inhouse_hemizygous: Option<i32>,

    /// Maximal frequency in HelixMtDb
    pub helixmtdb_frequency: Option<f32>,
    /// Maximal number of heterozygous carriers in HelixMtDb.
    pub helixmtdb_heterozygous: Option<i32>,
    /// Maximal number of homozygous carriers in HelixMtDb.
    pub helixmtdb_homozygous: Option<i32>,
}

impl Default for CaseQuery {
    /// Returns default values for a `CaseQuery` which makes all variants pass.
    fn default() -> Self {
        Self {
            effects: VariantEffect::all(),
            gnomad_exomes_enabled: Default::default(),
            gnomad_genomes_enabled: Default::default(),
            inhouse_enabled: Default::default(),
            helixmtdb_enabled: Default::default(),
            quality: Default::default(),
            genotype: Default::default(),
            selected_variants: Default::default(),
            transcripts_coding: true,
            transcripts_noncoding: true,
            var_type_snv: true,
            var_type_indel: true,
            var_type_mnv: true,
            max_exon_dist: Default::default(),
            gene_allowlist: Default::default(),
            genomic_regions: Default::default(),
            require_in_clinvar: Default::default(),
            clinvar_include_benign: true,
            clinvar_include_pathogenic: true,
            clinvar_include_likely_benign: true,
            clinvar_include_likely_pathogenic: true,
            clinvar_include_uncertain_significance: true,
            gnomad_exomes_frequency: Default::default(),
            gnomad_exomes_heterozygous: Default::default(),
            gnomad_exomes_homozygous: Default::default(),
            gnomad_exomes_hemizygous: Default::default(),
            gnomad_genomes_frequency: Default::default(),
            gnomad_genomes_heterozygous: Default::default(),
            gnomad_genomes_homozygous: Default::default(),
            gnomad_genomes_hemizygous: Default::default(),
            inhouse_carriers: Default::default(),
            inhouse_heterozygous: Default::default(),
            inhouse_homozygous: Default::default(),
            inhouse_hemizygous: Default::default(),
            helixmtdb_frequency: Default::default(),
            helixmtdb_heterozygous: Default::default(),
            helixmtdb_homozygous: Default::default(),
        }
    }
}

/// Information on the call as combined by the annotator.
#[derive(serde::Serialize, serde::Deserialize, PartialEq, Debug, Clone, Default)]
pub struct CallInfo {
    /// The genotype, if applicable, e.g., "0/1", "./1", "."
    pub genotype: Option<String>,
    /// Genotype quality score, if applicable
    pub quality: Option<f32>,
}

/// Definition of a sequence variant with per-sample genotype calls.
///
/// This uses a subset/specialization of what is described by the VCF standard
/// for the purpose of running SV queries in `varfish-server-worker`.
#[derive(serde::Serialize, serde::Deserialize, PartialEq, Debug, Clone)]
pub struct SequenceVariant {
    /// Chromosome name
    pub chrom: String,
    /// 1-based start position of the variant (or position on first chromosome
    /// for break-ends)
    pub pos: i32,
    /// Reference allele.
    pub reference: String,
    /// Alternative allele.
    pub alternative: String,

    /// Mapping of sample to genotype information for the SV.
    pub call_info: indexmap::IndexMap<String, CallInfo>,
}

impl SequenceVariant {
    /// Convert from VCF record.
    pub fn from_vcf(record: &vcf::Record, header: &vcf::Header) -> Result<Self, anyhow::Error> {
        let chrom = record.chromosome().to_string();
        let pos: usize = record.position().into();
        let pos = pos as i32;

        let reference = record.reference_bases().to_string();
        let alternative = record.alternate_bases()[0].to_string();

        let call_info = Self::build_call_info(record, header)?;

        Ok(Self {
            chrom,
            pos,
            call_info,
            reference,
            alternative,
        })
    }

    /// Build call information.
    fn build_call_info(
        record: &vcf::Record,
        header: &vcf::Header,
    ) -> Result<indexmap::IndexMap<String, CallInfo>, anyhow::Error> {
        let mut result = indexmap::IndexMap::new();

        for (name, sample) in header
            .sample_names()
            .iter()
            .zip(record.genotypes().values())
        {
            let genotype = if let Some(Some(vcf::record::genotypes::sample::Value::String(gt))) =
                sample.get(&vcf::record::genotypes::keys::key::GENOTYPE)
            {
                Some(gt.clone())
            } else {
                None
            };
            let quality = if let Some(Some(vcf::record::genotypes::sample::Value::Float(quality))) =
                sample.get(&vcf::record::genotypes::keys::key::CONDITIONAL_GENOTYPE_QUALITY)
            {
                Some(*quality)
            } else {
                None
            };

            result.insert(name.clone(), CallInfo { genotype, quality });
        }

        Ok(result)
    }
}

#[cfg(test)]
pub mod test {
    use rstest::rstest;

    #[rstest]
    #[case("tests/seqvars/query/empty.json")]
    #[case("tests/seqvars/query/full.json")]
    #[case("tests/seqvars/query/with_extra.json")]
    pub fn smoke_test_load(#[case] path: &str) -> Result<(), anyhow::Error> {
        let query: super::CaseQuery = serde_json::from_reader(std::fs::File::open(path)?)?;

        insta::assert_yaml_snapshot!(&query);

        Ok(())
    }
}
