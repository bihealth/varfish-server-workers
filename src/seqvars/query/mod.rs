//! Code implementing the "seqvars query" sub command.

mod interpreter;
mod schema;

use std::time::Instant;

use clap::{command, Parser};
use noodles_vcf as vcf;

use mehari::{annotate::seqvars::CHROM_TO_CHROM_NO, common::open_read_maybe_gz};
use rand_core::{RngCore, SeedableRng};
use thousands::Separable;
use uuid::Uuid;

use crate::{common::trace_rss_now, common::GenomeRelease};

use self::schema::SequenceVariant;

/// Command line arguments for `seqvars query` sub command.
#[derive(Parser, Debug)]
#[command(author, version, about = "Run query for seqvars", long_about = None)]
pub struct Args {
    /// Genome release to assume.
    #[arg(long, value_enum)]
    pub genome_release: GenomeRelease,
    /// Result set ID.
    #[arg(long)]
    pub result_set_id: Option<String>,
    /// Path to worker database to use for querying.
    #[arg(long)]
    pub path_db: String,
    /// Path to query JSON file.
    #[arg(long)]
    pub path_query_json: String,
    /// Path to input TSV file.
    #[arg(long)]
    pub path_input: String,
    /// Path to the output TSV file.
    #[arg(long)]
    pub path_output: String,

    /// Optional maximal number of total records to write out.
    #[arg(long)]
    pub max_results: Option<usize>,
    /// Optional seed for RNG.
    #[arg(long)]
    pub rng_seed: Option<u64>,
    /// Maximal distance to TAD to consider (unused, but required when loading database).
    #[arg(long, default_value_t = 10_000)]
    pub max_tad_distance: i32,
}

/// The structured result information of the result record.
#[derive(Debug, Default, serde::Serialize, serde::Deserialize)]
struct ResultPayload {}

/// A result record from the query.
#[derive(Debug, Default, serde::Serialize, serde::Deserialize)]
struct ResultRecord {
    sodar_uuid: Uuid,
    release: String,
    chromosome: String,
    chromosome_no: i32,
    reference: String,
    alternative: String,
    bin: u32,
    start: i32,
    end: i32,
    smallvariantqueryresultset_id: String,
    payload: String,
}

/// Utility struct to store statistics about counts.
#[derive(Debug, Default)]
struct QueryStats {
    pub count_passed: usize,
    pub count_total: usize,
    pub by_effect: indexmap::IndexMap<schema::VariantEffect, usize>,
}

/// Run the `args.path_input` VCF file and run through the given `interpreter` writing to
/// `args.path_output`.
fn run_query(
    interpreter: &interpreter::QueryInterpreter,
    args: &Args,
    _dbs: &crate::strucvars::query::Databases,
    rng: &mut rand::rngs::StdRng,
) -> Result<QueryStats, anyhow::Error> {
    let chrom_to_chrom_no = &CHROM_TO_CHROM_NO;
    let mut stats = QueryStats::default();

    // Open VCF file, create reader, and read header.
    let mut input_reader = open_read_maybe_gz(&args.path_input).map_err(|e| {
        anyhow::anyhow!("could not open file {} for reading: {}", args.path_input, e)
    })?;
    let mut input_reader = vcf::Reader::new(&mut input_reader);
    let input_header = input_reader.read_header()?;

    // Create output TSV writer.
    let mut csv_writer = csv::WriterBuilder::new()
        .has_headers(true)
        .delimiter(b'\t')
        .quote_style(csv::QuoteStyle::Never)
        .from_path(&args.path_output)?;

    // Read through input records using the query interpreter as a filter
    for input_record in input_reader.records(&input_header) {
        stats.count_total += 1;
        let record_seqvar = SequenceVariant::from_vcf(&input_record?, &input_header)
            .map_err(|e| anyhow::anyhow!("could not parse VCF record: {}", e))?;
        tracing::debug!("processing record {:?}", record_seqvar);

        let passes = interpreter.passes(&record_seqvar)?;

        let start = record_seqvar.pos;
        let end = start + record_seqvar.reference.len() as i32 - 1;
        let bin = mehari::annotate::seqvars::binning::bin_from_range(start - 1, end)? as u32;

        let result_payload = ResultPayload::default();

        let SequenceVariant {
            chrom: chromosome,
            reference,
            alternative,
            ..
        } = record_seqvar;
        let chromosome_no = *chrom_to_chrom_no
            .get(&chromosome)
            .expect("invalid chromosome") as i32;

        if passes.pass_all {
            // Finally, write out the record.
            let mut uuid_buf = [0u8; 16];
            rng.fill_bytes(&mut uuid_buf);
            csv_writer
                .serialize(&ResultRecord {
                    smallvariantqueryresultset_id: args.result_set_id.clone().unwrap_or(".".into()),
                    sodar_uuid: Uuid::from_bytes(uuid_buf),
                    release: match args.genome_release {
                        GenomeRelease::Grch37 => "GRCh37".into(),
                        GenomeRelease::Grch38 => "GRCh38".into(),
                    },
                    chromosome,
                    chromosome_no,
                    start,
                    end,
                    bin,
                    reference,
                    alternative,
                    payload: serde_json::to_string(&result_payload)
                        .map_err(|e| anyhow::anyhow!("could not serialize payload: {}", e))?,
                })
                .map_err(|e| anyhow::anyhow!("could not write record: {}", e))?;
        }
    }

    Ok(stats)
}

/// Main entry point for `seqvars query` sub command.
pub fn run(args_common: &crate::common::Args, args: &Args) -> Result<(), anyhow::Error> {
    let before_anything = Instant::now();
    tracing::info!("args_common = {:?}", &args_common);
    tracing::info!("args = {:?}", &args);

    // Initialize the random number generator from command line seed if given or local entropy
    // source.
    let mut rng = if let Some(rng_seed) = args.rng_seed {
        rand::rngs::StdRng::seed_from_u64(rng_seed)
    } else {
        rand::rngs::StdRng::from_entropy()
    };

    tracing::info!("Loading query...");
    let query: schema::CaseQuery =
        serde_json::from_reader(std::fs::File::open(&args.path_query_json)?)?;
    tracing::info!(
        "... done loading query = {}",
        &serde_json::to_string(&query)?
    );

    tracing::info!("Loading worker databases...");
    let before_loading = Instant::now();
    let path_worker_db = format!("{}/worker", &args.path_db);
    let dbs = crate::strucvars::query::load_databases(
        &path_worker_db,
        args.genome_release,
        args.max_tad_distance,
    )?;
    tracing::info!(
        "...done loading databases in {:?}",
        before_loading.elapsed()
    );

    trace_rss_now();

    tracing::info!("Translating gene allow list...");
    let hgvs_allowlist = if let Some(gene_allowlist) = &query.gene_allowlist {
        if gene_allowlist.is_empty() {
            None
        } else {
            Some(crate::strucvars::query::translate_gene_allowlist(
                gene_allowlist,
                &dbs,
            ))
        }
    } else {
        None
    };

    tracing::info!("Running queries...");
    let before_query = Instant::now();
    let query_stats = run_query(
        &interpreter::QueryInterpreter::new(query, hgvs_allowlist),
        args,
        &dbs,
        &mut rng,
    )?;
    tracing::info!("... done running query in {:?}", before_query.elapsed());
    tracing::info!(
        "summary: {} records passed out of {}",
        query_stats.count_passed.separate_with_commas(),
        query_stats.count_total.separate_with_commas()
    );
    tracing::info!("passing records by effect type");
    for (effect, count) in query_stats.by_effect.iter() {
        tracing::info!("{:?} -- {}", effect, count);
    }

    trace_rss_now();

    tracing::info!(
        "All of `seqvars query` completed in {:?}",
        before_anything.elapsed()
    );
    Ok(())
}

#[cfg(test)]
mod test {
    #[tracing_test::traced_test]
    #[rstest::rstest]
    #[case("tests/seqvars/query/Case_1.ingested.vcf")]
    #[case("tests/seqvars/query/dragen.ingested.vcf")]
    fn smoke_test(#[case] path_input: &str) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", path_input.split('/').last().unwrap());

        let tmpdir = temp_testdir::TempDir::default();
        let path_output = format!("{}/out.tsv", tmpdir.to_string_lossy());
        let path_input: String = path_input.into();
        let path_query_json = path_input.replace(".ingested.vcf", ".query.json");

        let args_common = Default::default();
        let args = super::Args {
            genome_release: crate::common::GenomeRelease::Grch37,
            path_db: "tests/seqvars/query/db".into(),
            path_query_json,
            path_input,
            path_output: path_output,
            max_results: None,
            rng_seed: Some(42),
            max_tad_distance: 10_000,
            result_set_id: None,
        };
        super::run(&args_common, &args)?;

        insta::assert_snapshot!(std::fs::read_to_string(args.path_output.as_str())?);

        Ok(())
    }
}
