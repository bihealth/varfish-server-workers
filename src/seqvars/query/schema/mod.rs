//! Internal Rust representations for query and data.

pub mod data;
pub mod query;
