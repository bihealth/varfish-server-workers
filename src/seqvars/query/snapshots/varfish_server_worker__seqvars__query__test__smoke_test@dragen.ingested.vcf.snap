---
source: src/seqvars/query/mod.rs
expression: "std::fs::read_to_string(args.path_output.as_str())?"
---
sodar_uuid	release	chromosome	chromosome_no	reference	alternative	bin	start	end	smallvariantqueryresultset_id	payload
a2242722-6377-cc86-7d51-ad3f130af08a	GRCh37	17	17	CA	C	899	41256074	41256075	.	{}
d13451de-7160-efa2-b230-76fd782de967	GRCh37	MT	25	A	G	585	750	750	.	{}

